### PR TITLE
POC: allow overriding parts of rule.yml based on compiled product

### PR DESCRIPTION
**Pull request by vojtapolasek**
This is just POC and work in progress.

#### Description:

If you place a file into <some_rule_folder>/rule_overrides and call it like <prodtype>.yaml, where prodtype is an existing prodtype, it will override some keys of the rule.yaml file.
Currently, the set of overridable keys is limited to documentation_complete, title, description, severity, ocil and ocil_clause.

#### Rationale:

It can make easier to produce different rule verbiage for different products. The PR #10550 might make use of this feature I believe.

#### Review Hints:

- add a new folder rule_overrides into some rule (e.g. linux_os/guide/services/ssh/file_groupownership_ssh_private_keys.
- create a file called rhel9.yml in the folder
- insert 
```
description: "blahblah"
```
into the file.
- Build rhel8 and rhel9 product.
- Examine compiled rules for both products and you should see the difference. E.g.
```
diff -u build/rhel8/rules/file_groupownership_ssh_private_keys.yml build/rhel9/rules/file_groupownership_ssh_private_keys.yml
```

#### Needs to be done

- [x] tests
- [ ] documentation

### CODE REVIEW

File: ssg/build_yaml.py [hunk 1]
```diff
@@ -849,6 +857,34 @@ def load_policy_specific_content(self, rule_filename, env_yaml):
...
+        applicable_filename = env_yaml["product"] + ".yml"
+        return os.path.basename(filename) == applicable_filename
+
+    def apply_override_files(self, rule_file, env_yaml):
+        rule_dir = os.path.dirname(rule_file)
+        overrides = self.get_rule_override_files(rule_dir)
+        applicable_files = [
+            f for f in overrides if self.check_if_override_is_applicable(f, env_yaml)]
+        if len(applicable_files) == 0:
+            return
```
[the remark this answers is not in the document]
Comment (vojtapolasek) [reply ssg/build_yaml.py:882]: I think it is actually not possible to override the rule with multiple files. The file name is strictly defined; <prodtype>.yml.

**Comment by vojtapolasek** [comment]
Hello @lbragstad  what do you think about this change? I think it might do what you want in https://github.com/ComplianceAsCode/content/pull/10550
What keys found in rule.yml would you like to override?

**Comment by vojtapolasek** [comment]
The automatus test failures are irrelevant I believe. there is no new rule added / altered.

**Comment by ggbecker** [comment]
I find the need of a `rule_overrides` folder a bit strange. Is there any reason why not to enable overriding the rules with files within the same directory as the `rule.yml` itself. In remediations we don't need any extra directory for example when overriding the remediation.

Maybe there is a reason I'm not seeing here.

**Comment by vojtapolasek** [comment]
@ggbecker I think the only reason I have for this is the fact that the "rule root" directory is already quite cluttered. It can have already 5 subdirectories or even more. But I am not forcing this, I just thought it will be cleaner.

**Comment by ggbecker** [comment]
> @ggbecker I think the only reason I have for this is the fact that the "rule root" directory is already quite cluttered. It can have already 5 subdirectories or even more. But I am not forcing this, I just thought it will be cleaner.

I see your point. I just though it would differ from the behavior we have for OVAL/Remediation where you can put a file in the same folder. Either is fine by me.

**Comment by vojtapolasek** [comment]
Hello @lbragstad  @jhrozek  any thoughts on this mechanism?

File: ssg/build_yaml.py [hunk 2]
```diff
@@ -691,6 +691,13 @@ class Rule(XCCDFEntity, Templatable):
         "severity",
     }
 
+    OVERRIDABLE_KEYS = {
+        "documentation_complete",
+        "title",
+        "description",
```
[the remark this answers is not in the document]
Comment (vojtapolasek) [reply ssg/build_yaml.py:697]: Sure, this is just a POC, I am waiting for feedback as shown in previous comments. AFter we settle down on specific things, I will add documentation.

File: docs/manual/developer/06_contributing_with_content.md [hunk 3]
```diff
@@ -418,6 +420,38 @@ that begin with underscores are not meant to be used in descriptions.
...
+
+- title
+
+- description
+
+- rationale
+
+- ocil
+
+- ocil_clause
```
Comment (Vincent056) [inline docs/manual/developer/06_contributing_with_content.md:448]: I wonder if warnings can also be in this list

Comment (vojtapolasek) [reply docs/manual/developer/06_contributing_with_content.md:448]: @Vincent056  I added support for warnings.

**Comment by vojtapolasek** [comment]
Note that content test filtering is detecting "overrides" as a change in rule for some reason. We will need to investigate it before merging this PR.

**Comment by vojtapolasek** [comment]
Hello @yuumasato, do you have any feedback for me regarding this PR?

**Comment by yuumasato** [comment]
@vojtapolasek I don't have any feedback right now.
We have not looked at how the support for multiple kubernetes CIS benchmarks can or should look like.

I guess we could close this PR for the time being and reopen when we start to look into this again.
Thoughts on this @rhmdnd @Vincent056 ?

**Comment by vojtapolasek** [comment]
Closing now, feel free to reopen.

**Pull request closed without merging**